### PR TITLE
Tracking NN training [WIP]

### LINKED ISSUE #125: Real-time tracking while fitting NN/deep learning
If I'm training a model and allowing tracking, I should be able to open a new terminal and query the results in real-time (or at least see some of the parameters collected). This can happen every epoch on the network.

### PR DESCRIPTION
**Pull request by neelasha23**
## Issue ticket number and link
Closes #125

## Checklist before requesting a review

- [ ] I have performed a self-review of my code
- [ ] I have added thorough [tests](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#testing) (when necessary).
- [ ] I have added the right documentation in the [docstring](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#documenting-changes-and-new-features) and [changelog](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#changelog) (when needed)

### CODE REVIEW

File: docs/comparison/nn_tracking.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+history = model.fit(X_train, y_train, epochs=no_epochs, batch_size=len(X_train),  verbose=0)
+>>>>>>> fe9a4c5 (verbose)
+```
+
+```{code-cell} ipython3
+model.summary()
+```
+
+```{code-cell} ipython3
+<<<<<<< HEAD
```
Comment (idomic) [inline docs/comparison/nn_tracking.md:92]: I think you might have a conflict here

**Comment by neelasha23** [comment]
Facing timeout issues in this particular PR only. Tensorflow installation is also timing out.
@edublancas Can you have a look at the code and let me know if this is in the right direction?

Ideas that I had in mind to take this forward:
1. Demonstrate using multiple models, like different combinations of activation functions, layers(Dense, Conv, etc), number of hidden units etc
2. Select a particular usecase like text/image classification.
3. Plot epoch wise validation loss and accuracy
4. I'm not sure if layer wise weights needs to be tracked?

**Comment by edublancas** [comment]
> Facing timeout issues in this particular PR only (have done a rebase on master). Tensorflow installation is also timing out.

add tensorflow and keras to the documentation requirements so they're installed in the setup step instead of when running the notebook

> Can you have a look at the code and let me know if this is in the right direction?

I took a look at the code, looks like you're training the NN and storing the final metrics (?), the idea here is to produce multiple metrics and store them on each X epochs to see how they improve as we train the NN for more epochs, the end output should be a plot with epochs in the X axis and the metrics in the Y axis. I'm unsure what's the best way to store the data since this is a new use case, so you might want to experiment a bit. If needed, we can make changes to the SQLiteTracker to simplify the user experience.

> Ideas that I had in mind to take this forward:

I think let's keep it simple, we want to quickly demonstrate the tracking abilities. so use a small dataset and a small network so it's fast to train. MNIST with a small network is probably the best option.

**Comment by neelasha23** [comment]
Ok. I think I understood it wrong then. Will experiment more with the epochs and metrics

**Comment by neelasha23** [comment]
I took a look at the code again. `_ = experiment.log_dict(history.history)` : this is storing the epoch wise validation loss and accuracy. the docs is up now so you can have a look. 
When you say we need to store epoch wise metrics do you mean something else?

I agree the way it's being stored right now can be improved. Like maybe 
```
{"epoch 1" : { "loss":0.678, "acc": 0.5224, .. any other metrics},
 "epoch 2" : {....}
}
```

So there'll be one plot per metric and it would be the trend of the metric across all epochs.

Please confirm if my understanding is correct.
@edublancas

**Comment by idomic** [comment]
Check [this one](https://medium.com/analytics-vidhya/tracking-deep-learning-experiments-using-keras-mlflow-and-mongodb-732fc771266c) fig. 17.
Accuracy and loss is great, see if you can track anything else that's in there.
Also, at this point let's try to solve this without updating the API, there are ways in the docs to do it (for instance, manual tracking).

We should also before merging, add a flag to the CI to skip executing this notebook as we're planning to push it to 3rd party (kaggle).

**Comment by neelasha23** [comment]
I've added the MNIST dataset and plots. The doc build is failing on the model fit cell (checking this).
I'm yet to add other metrics. I tried precision, recall etc but these work with binary classification. I think the data needs to be transformed for these to work (checking the blog shared by @idomic )

I have a doubt on the plots:
Currently it's taking the metrics from the history variable but it should be queried from the tracker right?
 Also, from the tracker guide I see this `experiment.log_confusion_matrix(y_test, y_pred)` which internally calls the ConfusionMatrix API. Do we need something similar here? Like an API `NeuralNetworkMetrics` that would plot metrics vs epochs. And a corresponding `log_nn_metrics` function?

@edublancas

**Comment by neelasha23** [comment]
As discussed have added the following:
1. real-time reading of training metrics using callback
2. logging the metrics in a tracker (single experiment) using `upsert`
3. Demonstration of querying the tracker in realtime

Currently dataset has 500 images only. Once this is approved I'll remove the data limits.
And add this notebook to `exclude_pattern` in `docs/conf.py`

@idomic

**Comment by edublancas** [comment]
is it ready for review? if so, mark it as ready and request my review

**Comment by neelasha23** [comment]
> is it ready for review? if so, mark it as ready and request my review

done

File: docs/comparison/nn_tracking.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.14.1
+kernelspec:
+  display_name: Python 3 (ipykernel)
+  language: python
+  name: python3
+---
+
+# Real-time tracking of Neural Network training
```
Comment (edublancas) [inline docs/comparison/nn_tracking.md:14]: let's make this shorter:

```suggestion
# Real-time tracking
```

File: docs/comparison/nn_tracking.md [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+SQLiteTracker provides a powerful and flexible way to track computational (e.g., Machine Learning) experiments using a SQLite database.
+
+In this tutorial we demonstrate training a small network on the Fashion MNIST dataset and tracking the training metrics in realtime. We would also see how you can query the tracked metrics while training is going on, and visualise the metrics vs epoch plots.
```
Comment (edublancas) [inline docs/comparison/nn_tracking.md:18]:
```suggestion
This tutorial demonstrates training a small network on the Fashion MNIST dataset and tracking the training and validation metrics in real-time. We would also see how you can query the tracked metrics while training is ongoing and visualize the metrics vs. epoch plots.
```

File: docs/comparison/nn_tracking.md [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+## Create the experiment tracker
+
+```{code-cell} ipython3
+from pathlib import Path
+from sklearn_evaluation import SQLiteTracker
+
+db = Path('nn_experiments.db')
+
+if db.exists():
+    db.unlink()
```
Comment (edublancas) [inline docs/comparison/nn_tracking.md:31]: let's put this in a cell above and add the `:hide-cell:` tag, so it doesn't show by default https://jupyterbook.org/en/stable/interactive/hiding.html#hide-entire-code-cells

File: docs/comparison/nn_tracking.md [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+                    train_labels, 
+                    validation_data=(val_images, val_labels), 
+                    epochs=epoch_count, 
+                    verbose=0, 
+                    callbacks=[TrackLossandAccuracyCallback()])
+```
+
+## Track metrics while training
+
+While the training is going on, you may visualise the metrics plot by opening another terminal / notebook and run the below steps.
```
Comment (edublancas) [inline docs/comparison/nn_tracking.md:136]:
```suggestion
While the training is ongoing, you may visualize the metrics plot by opening another terminal/notebook and running the steps below.
```

File: docs/comparison/nn_tracking.md [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    uuid,
+    json_extract(parameters, '$.accuracy') as accuracy,
+    json_extract(parameters, '$.loss') as loss,
+    json_extract(parameters, '$.val_accuracy') as val_accuracy,
+    json_extract(parameters, '$.val_loss') as val_loss
+    FROM experiments
+""")
+```
+
+Extract and plot relevant metrics against epcohs
```
Comment (edublancas) [inline docs/comparison/nn_tracking.md:154]:
```suggestion
Extract and plot relevant metrics against epochs:
```

**Review by edublancas — CHANGES_REQUESTED** [review]
looks great! added some comments. found some typos and style issues, please pass this through Grammarly

**Comment by neelasha23** [comment]
Resolved the comments and also checked with Grammarly.
Should I add this guide to `exclude_patterns` in `conf.py`
@edublancas

**Review by edublancas — APPROVED** [review]

**Comment by edublancas** [comment]
> Should I add this guide to exclude_patterns in conf.py

it's currently configured to run each time we build the docs, right? if so, let's leave it this way and exclude its execution only if it gives trouble down the road.
